client/src/components/Visualization/charts/Geochart.js

```
@@ -1,59 +1,67 @@
+    async function fetchStateTotals() {
+      /* 
+      dataTable is read by the Chart component into Geochart format.
+      The states are colored according to the value in the second column of
+        the table ("Total Ad Spend (USD)").
+      The first column (state name) and third column (info on total state ad 
+        spend and top advertiser) are shown in the tooltip.
+      */ 
+      let dataTable = [["State", "Total Ad Spend (USD)", {type: 'string', role: 'tooltip'}]];
 
-  // Firestore data converter converts snapshots to custom objects.
-  const adConverter = {
-    toFirestore(ad) {
-      return {
-        id: ad.id,
-      };
-    },
-    fromFirestore(snapshot, options) {
-      const data = snapshot.data(options);
-      return new GeochartAd(data.id);
-    },
-  };
+      let stateSpendPromises = [];
+      let topAdvertiserPromises = [];
+
+      for (let state of STATES) {
+        stateSpendPromises.push(database.collection(STATES_COLLECTION)
+            .doc(state.toLowerCase())
+            .get()
+        );
+        topAdvertiserPromises.push(database.collection(STATES_COLLECTION)
+            .doc(state.toLowerCase())
+            .collection('advertisers')
+            .orderBy('totalAdvertiserSpend')
+            .limit(1)
+            .get()
+        );
+      }
+
+      let stateResults = await Promise.all(stateSpendPromises);
+      let advertiserResults = await Promise.all(topAdvertiserPromises);
+      const zippedResults = stateResults.map((e, i) => [e,advertiserResults[i]]);
 
-  useEffect(async () => {
-    const data = [['State', 'Total Ad Spend Across All Advertisers']];
-    for (const state in states) {
-      data.push([
-        state,
-        Math.floor(Math.random() * Math.floor(250000) + 10000),
-      ]); // Update data table.
+      for (let result of zippedResults) {
+        let stateTotal = result[0];
+        let topAdvertiser = result[1];
+        if (stateTotal.data() !== undefined) {  
```

Under what circumstances would data be undefined?  Should you throw an error?